# Strategies for Flux Time Series Preprocessing
### From simple to advanced methods

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from itertools import chain
sns.set_style('whitegrid')
warnings.simplefilter('ignore', FutureWarning)

In [4]:
%run -i "load\train.py" #run script for loading

One of the first major challenges in this competition is to properly preprocess the light curve time series data. Unlike many other "well-behaved" time series, the light curves here are not only irregular in terms of observation intervals, but also unsynchronised across different light bands. This makes applying many existing tools and techniques difficult because they often require a regular time series, preferably without missing values. Let us first look at what we have:

In [5]:
train_series = training_set
train_metadata = training_set_metadata

In [6]:
train_series.head(10)

object_id         mjd  passband         flux   flux_err  detected
0        615  59750.4229         2  -544.810303   3.622952         1
1        615  59750.4306         1  -816.434326   5.553370         1
2        615  59750.4383         3  -471.385529   3.801213         1
3        615  59750.4450         4  -388.984985  11.395031         1
4        615  59752.4070         2  -681.858887   4.041204         1
5        615  59752.4147         1 -1061.457031   6.472994         1
6        615  59752.4224         3  -524.954590   3.552751         1
7        615  59752.4334         4  -393.480225   3.599346         1
8        615  59752.4435         5  -355.886780  10.421921         1
9        615  59767.2968         2  -548.013550   3.462291         1

There are actually 3 time series (flux, flux_err, detected) per band per object. However, as a first step, here we will only focus on the "flux" time series. Even so, we still have 6 passbands for each object.

We first look at the distribution of time series lengths.

In [ ]:
ts_lens = train_series.groupby(['object_id', 'passband']).size()
f, ax = plt.subplots(figsize=(12, 6))
sns.distplot(ts_lens, ax=ax)
ax.set_title('distribution of time series lengths')
plt.show()

As we see here, the lengths of light curve time series have a large range, so we may face difficulties trying to convert them all into the same length.

Let us also look at the times at which observations happen, and count the number of observations made at each point in time:

In [ ]:
f, ax = plt.subplots(figsize=(12, 6))
sns.distplot(train_series['mjd'], ax=ax, bins=200)
ax.set_title('number of observations made at each time point')
plt.show()

What we see here shows that the observations are not taken evenly through the entire sampling period. There are periods where more samples than usual are taken, but also periods with few observations.

As for each individual object:

In [ ]:
f, ax = plt.subplots(figsize=(12, 6))
sns.distplot(train_series[train_series['object_id'] == 615]['mjd'], ax=ax, bins=200)
ax.set_title('number of observations made at each time point')
plt.show()

What about for a single light band?

In [ ]:
f, ax = plt.subplots(figsize=(12, 6))
sns.distplot(train_series[(train_series['object_id'] == 615) 
                          & (train_series['passband'] == 2)]['mjd'], ax=ax, bins=200)
ax.set_title('number of observations made at each time point')
plt.show()

We see that even for individual objects, observations are not taken evenly. There can be large gaps between periods of rich observations.

Let us also look at the number of observations for each object:

In [ ]:
obj_obs_count = train_series['object_id'].value_counts().reset_index()
obj_obs_count.columns = ['object_id', 'count']

obj_obs_count_w_ddf = pd.merge(
    obj_obs_count, train_metadata[['object_id', 'ddf']], on='object_id')

selected = obj_obs_count_w_ddf.groupby('ddf')['count'].value_counts()
selected.index.names = ['ddf', 'count_val']
selected = selected.reset_index().pivot('count_val', 'ddf',
                                        'count').rename(columns={
                                            0: 'nonddf',
                                            1: 'ddf'
                                        }).fillna(0)
selected['total'] = selected['nonddf'] + selected['ddf']

f, ax = plt.subplots(figsize=(12, 6))
ax.vlines(x=selected.index, ymin=0, ymax=selected['total'], 
          colors='red', label='ddf')
ax.vlines(x=selected.index, ymin=0, ymax=selected['nonddf'], 
          colors='blue', label='non-ddf')
ax.set_title('number of observations per object')
plt.legend()
plt.show()

As can be seen above, there is a large gap between the number of observations for an average DDF object and non-DDF object.

Judging by the observations we made so far, the processing of the light curve time series is indeed a challenging task. We have the following difficulties to overcome:
1. The observations, even for one object in one band, are not evenly spaced, so adjacent observations in the time series are not equally "close" to each other.
2. The length of the time series vary quite a lot, even among the colour bands of the same object. At no point do we have multiple colour band observations simultaneously, so we cannot naively treat the multiple light bands' time series as a multivariate series.
3. There are some large gaps between periods where data is avaiable for an object, so we may miss characteristics of the time series at a certain period if we only look at global features.
4. The time series for different objects are not synchronised, making comparisons between objects difficult. We will have to disocover features that are invariant with time shifting or slight stretching.

#### Ignore time values

There are multiple strategies to approach this problem, with varying degrees of complexity.
* We may completely ignore the time values and simply treat the time series as sequences. Many sequence features such as min, max, range, std, monotonicity etc. are invariant with the loss of time interval information, so we can still build many useful features. However, it comes with two major disadvantages:
    1. We will not able to align observations that occur close together, so we will have to analyse the time series for each passband individually without exploiting their relationships.
    2. We will lose access to many useful feature extraction techniques, or at least the form of these techniques that are typically implemented in common tools if we throw out time interval information. Features like autocorrelation do not make sense if we do not know the actual time gaps between observations.
    
Performing time-less analysis is simple. We do not even need much preprocessing, which makes it a good starting point. In the following example, we extract four very basic features from the time-less time series:

In [ ]:
simple_features = train_series.groupby(
    ['object_id', 'passband'])['flux'].agg(
    ['mean', 'max', 'min', 'std']).unstack('passband')
simple_features.head().T

Even if we cannot extract interaction features from multiple bands directly, we may still be able to discover their relationships from the features we have in downstream analysis and learning tasks.

**What not to do here:** grab cesium or tsfresh, spin up a few cores and extract all the fancy features you can get. Be aware that many of the features in these libraries do not work well with unevenly-spaced time series. I wasted many CPU hours here myself :(

However, we cannot simply ignore the rich information we just threw out. The time-less approach won't even allow us to discover periodicity! Plus, we know from some astronomy background that it is the composition of light at different frequencies (and by extension, bands) that allow us to identify objects in the first place. We really do not want to ignore the relationship between passbands. So we can come up with a different approach:

#### Convert to regular TS

* We may take all **mjd** values in the dataset and declare them 'valid observation timestamps', then construct a time series for each object and each passband that include all these observation timestamps, filling in NA if obseration is not available. Now that all objects and all passband have time series of the same length, they are comparable, and many time series analysis techniques can be applied as long as they can deal with missing values. One immediate drawback of this method is that none of the objects will have multiple passband data at a given timestamp, and there will be a huge number of NAs. The time series will also be really long due to the large number of possible timestamps. Also, we will face more issues when we want to build a model and apply it to new data, as the new timestamps will not match the timestamps in our current dataset. We can make a compromise by binning **mjd** and averaging the observations within a bin, and use all bins within the range of **mjd** as possible time steps in the time series. By doing so, we will be able to obtain a collection of time series that are equally spaced, not overwhemingly NA, and have the same length.

In [ ]:
print(f'mjd unique values: {train_series["mjd"].nunique()}')

In [ ]:
print(f'int (1day) mjd unique values: {train_series["mjd"].astype(int).nunique()}')
print(f'''int mjd bins: {train_series["mjd"].astype(int).max()
      - train_series["mjd"].astype(int).min() + 1}''')

In [ ]:
print(f'5day mjd unique values: {(train_series["mjd"]/5).astype(int).nunique()}')
print(f'''5day mjd bins: {int((train_series["mjd"].astype(int).max()
      - train_series["mjd"].astype(int).min()) / 5 + 1)}''')

As we see above, while using each unique **mjd** value as time steps is really bad, binning by day or even by 5 days greatly reduces time series length. Here is an example of constructing the resulting time series using binned observations:

In [ ]:
# binning
ts_mod = train_series[['object_id', 'mjd', 'passband', 'flux']].copy()
ts_mod['mjd_d5'] = (ts_mod['mjd'] / 5).astype(int)
ts_mod = ts_mod.groupby(['object_id', 'mjd_d5', 'passband'])['flux'].mean().reset_index()
ts_mod.head(10)

In [ ]:
# pivotting
ts_piv = pd.pivot_table(ts_mod, 
                        index='object_id', 
                        columns=['mjd_d5', 'passband'], 
                        values='flux',
                        dropna=False)
ts_piv.head(10)

In [ ]:
# resetting column index to fill mjd_d5 gaps 
t_min, t_max = ts_piv.columns.levels[0].min(), ts_piv.columns.levels[0].max()
t_range = range(t_min, t_max + 1)
mux = pd.MultiIndex.from_product([list(t_range), list(range(6))], 
                                 names=['mjd_d5', 'passband'])
ts_piv = ts_piv.reindex(columns=mux).stack('passband')
ts_piv.head(10)

We have now converted the unaligned time series into aligned, equally spaced time series using binning-and-averaging. We can see below that despite our efforts, nearly 90% of the values in our new time series are NaN, so there will be major challenges trying to make use of the converted data with tools that cannot cope well with NAs. Normal procedures like mean imputation or interpolation won't work well here due to the prevalance of missing values. However, some algorithms like LSTM may learn to ignore padded values and data gaps and still manage to extract valuable information from the time series. It is even possible to build multiple time series collections with different bins to look for features at different time scales.

In [ ]:
np.mean(np.ravel(pd.isna(ts_piv).values))

One of the drawbacks of the method above is that binning appears to be quite arbitrary. What we are effectively doing is sampling at different time points and averaging flux values around these points. So why not do it properly?

#### Sample with a distance kernel

* We may sample the time series data at evenly-spaced time steps, using a time kernel to determine how much weight we put on each observation when taking the moving average.

In [7]:
def time_kernel(diff, tau):
    return np.exp(-diff ** 2 / (2 * tau ** 2))

In [8]:
t_min, t_max = train_series['mjd'].min(), train_series['mjd'].max()
t_min, t_max

(59580.0343, 60674.363)

In [9]:
sample_points = np.array(np.arange(t_min, t_max, 20))
sample_points

array([59580.0343, 59600.0343, 59620.0343, 59640.0343, 59660.0343,
       59680.0343, 59700.0343, 59720.0343, 59740.0343, 59760.0343,
       59780.0343, 59800.0343, 59820.0343, 59840.0343, 59860.0343,
       59880.0343, 59900.0343, 59920.0343, 59940.0343, 59960.0343,
       59980.0343, 60000.0343, 60020.0343, 60040.0343, 60060.0343,
       60080.0343, 60100.0343, 60120.0343, 60140.0343, 60160.0343,
       60180.0343, 60200.0343, 60220.0343, 60240.0343, 60260.0343,
       60280.0343, 60300.0343, 60320.0343, 60340.0343, 60360.0343,
       60380.0343, 60400.0343, 60420.0343, 60440.0343, 60460.0343,
       60480.0343, 60500.0343, 60520.0343, 60540.0343, 60560.0343,
       60580.0343, 60600.0343, 60620.0343, 60640.0343, 60660.0343])

In [10]:
weights = time_kernel(np.expand_dims(sample_points, 0) 
                      - np.expand_dims(train_series['mjd'].values, 1), 5)
ts_mod = train_series[['object_id', 'mjd', 'passband', 'flux']].copy()
for i in range(len(sample_points)):
    ts_mod[f'sw_{i}'] = weights[:, i]

In [11]:
ts_mod

object_id         mjd  passband         flux           sw_0  \
0              615  59750.4229         2  -544.810303  6.743158e-253   
1              615  59750.4306         1  -816.434326  6.398397e-253   
2              615  59750.4383         3  -471.385529  6.071248e-253   
3              615  59750.4450         4  -388.984985  5.800215e-253   
4              615  59752.4070         2  -681.858887  8.352701e-259   
5              615  59752.4147         1 -1061.457031  7.920806e-259   
6              615  59752.4224         3  -524.954590  7.511226e-259   
7              615  59752.4334         4  -393.480225  6.962548e-259   
8              615  59752.4435         5  -355.886780  6.494101e-259   
9              615  59767.2968         2  -548.013550  2.569150e-305   
10             615  59767.3045         1  -815.188599  2.425159e-305   
11             615  59767.3122         3  -475.516052  2.289233e-305   
12             615  59767.3233         4  -405.663818  2.106574e-305   
13             615  59767.3343         5  -421.199066  1.939933e-305   
14             615  59770.2179         2  -554.903198  6.809169e-315   
15             615  59770.2256         1  -820.042786  6.421761e-315   
16             615  59770.2334         3  -477.004730  6.051774e-315   
17             615  59770.2445         4  -400.270386  5.561682e-315   
18             615  59770.2557         5  -415.286896  5.107367e-315   
19             615  59779.3188         2  -630.523682   0.000000e+00   
20             615  59779.3265         1  -921.002502   0.000000e+00   
21             615  59779.3342         3  -518.533997   0.000000e+00   
22             615  59779.3452         4  -422.184509   0.000000e+00   
23             615  59779.3562         5  -422.815094   0.000000e+00   
24             615  59782.1897         2  -280.039520   0.000000e+00   
25             615  59782.1974         1  -449.095612   0.000000e+00   
26             615  59782.2051         3  -316.704865   0.000000e+00   
27             615  59782.2162         4  -332.885437   0.000000e+00   
28             615  59782.2274         5  -365.075775   0.000000e+00   
29             615  59797.2861         2   391.399231   0.000000e+00   
...            ...         ...       ...          ...            ...   
1421675  130779836  60464.2127         2     3.294472   0.000000e+00   
1421676  130779836  60466.2566         0   113.988297   0.000000e+00   
1421677  130779836  60490.1588         2    12.464979   0.000000e+00   
1421678  130779836  60491.1907         1   -13.734149   0.000000e+00   
1421679  130779836  60496.1248         2   -14.478923   0.000000e+00   
1421680  130779836  60497.1472         3    12.178056   0.000000e+00   
1421681  130779836  60498.1867         3   -20.815857   0.000000e+00   
1421682  130779836  60499.1367         0    54.361107   0.000000e+00   
1421683  130779836  60503.1145         3     9.912425   0.000000e+00   
1421684  130779836  60505.1075         2    17.719660   0.000000e+00   
1421685  130779836  60506.0962         5   -86.457382   0.000000e+00   
1421686  130779836  60510.2053         4   -39.215195   0.000000e+00   
1421687  130779836  60515.2731         5   -81.113083   0.000000e+00   
1421688  130779836  60524.0372         0   136.082718   0.000000e+00   
1421689  130779836  60525.0621         0    43.910740   0.000000e+00   
1421690  130779836  60526.1904         3    59.376686   0.000000e+00   
1421691  130779836  60528.0336         3     5.993599   0.000000e+00   
1421692  130779836  60530.0978         1     2.986908   0.000000e+00   
1421693  130779836  60536.0451         5     5.653964   0.000000e+00   
1421694  130779836  60537.1470         3   -41.850040   0.000000e+00   
1421695  130779836  60542.0489         4   -60.500492   0.000000e+00   
1421696  130779836  60543.0247         4   -48.527161   0.000000e+00   
1421697  130779836  60545.9844         5    32.006413   0.000000e+00   
1421698  130779836  60546.9804         5    68.15

In [12]:
def group_transform(chunk):
    sample_weights = chunk[[f'sw_{i}' for i in range(len(sample_points))]]
    sample_weights /= np.sum(sample_weights, axis=0)
    weighted_flux = np.expand_dims(chunk['flux'].values, 1) * sample_weights.fillna(0)
    return np.sum(weighted_flux, axis=0)

In [13]:
# only using a small sample as this step is slower than the other steps
ts_samp = ts_mod[ts_mod['object_id'].isin([615, 713])].groupby(
    ['object_id', 'passband']).apply(group_transform)

In [14]:
%%time
ts_samp_all = ts_mod.groupby(['object_id', 'passband']).apply(group_transform)

Wall time: 3min


In [31]:
ts_samp_all.index.levels[0]

Int64Index([      615,       713,       730,       745,      1124,      1227,
                 1598,      1632,      1920,      1926,
            ...
            130684460, 130695262, 130698059, 130716752, 130727624, 130739978,
            130755807, 130762946, 130772921, 130779836],
           dtype='int64', name='object_id', length=7848)

We can see above that we have managed to sample from the time series, using the time difference from the sample point to determine moving average weights.

There are still many issues with this method. For example, it still does not solve the problem of long gaps between available data, and the use of 0 as a filler value for NA is questionable (however we cannot retain NA, as that will make calculating moving average difficult). Overall, this is still a better approach than simple binning, and still relatively cheap to compute.

#### Use time / time difference as a feature

This idea is quite simple. Why not just concatenate the flux values with the timestamp at which they are observed, or the time elapsed since last observation, and pass the whole sequence to a learning algorithm? Indeed, some algorithms have been shown to be able to cope with that. This is not so different from "given (x, y), fit the curve and extract features from the curve". An LSTM or even a regular MLP might be able to deal with that. There is an [example](https://arxiv.org/pdf/1711.10609.pdf) of this being applied to a very similar problem.

Also, some time series feature extractors accept (t, x) pairs instead of (x) sequences.

(WIP)

#### Convert time series to phase series

If we know that some of the time series will be periodic, we can attempt to find the optimal period and convert the time series to phase domain, so that we can avoid the issues of observation gap, alleviate the problem of sparse observations and make better use of the periodic property. We may also assume that for most objects that exhibit period behaviour, the period is the same for all colour bands and the phases should all be in sync, because they are likely driven by the same physical event.
To discover the period of the time series, we first normalise all series then extract features about the most prominent frequencies / periods.

In [ ]:
groups = train_series.groupby(['object_id', 'passband'])

In [ ]:
def normalise(ts):
    return (ts - ts.mean()) / ts.std()

In [ ]:
times = groups.apply(
    lambda block: block['mjd'].values).reset_index().rename(columns={0: 'seq'})

In [ ]:
flux = groups.apply(
    lambda block: normalise(block['flux']).values
).reset_index().rename(columns={0: 'seq'})

In [ ]:
times_list = times.groupby('object_id').apply(lambda x: x['seq'].tolist()).tolist()
flux_list = flux.groupby('object_id').apply(lambda x: x['seq'].tolist()).tolist()

In [ ]:
import cesium.featurize as featurize
from scipy import signal
import warnings

To save some kernel time, we will only extract features from a small subset of objects.

([**FIXED - See below**]For some reason, the frequencies features of the cesium package are not working properly for me, so I have to revert to the method used by Michal Haltuf in his notebook [Feature extraction using period analysis](https://www.kaggle.com/rejpalcz/feature-extraction-using-period-analysis). It takes quite some time even for just 20 objects as you can see below:


In [ ]:
# def extract_freq(t, m, e):
#     fs = np.linspace(2*np.pi/0.1, 2*np.pi/500, 10000)
#     pgram = signal.lombscargle(t, m, fs, normalize=True)
#     return fs[np.argmax(pgram)]

# N = 20
# warnings.simplefilter('ignore', RuntimeWarning)
# feats = featurize.featurize_time_series(times=times_list[:N],
#                                         values=flux_list[:N],
#                                         features_to_use=['freq1'],
#                                         custom_functions={'freq1': extract_freq},
#                                         scheduler=None)

**EDIT:** I found why the features from the cesium package is acting weird. The frequency features extracted by cesium are oscilation frequencies (cycles per unit time) rather than angular frequencies (rad per unit time), so they differ by $2\pi$. We can now use cesium's 'freqN_freq' feature extractor to extract features slightly faster:

In [ ]:
warnings.simplefilter('ignore', RuntimeWarning)
N = 100
cfeats = featurize.featurize_time_series(times=times_list[:N],
                                        values=flux_list[:N],
                                        features_to_use=['freq1_freq',
                                                        'freq1_signif',
                                                        'freq1_amplitude1'],
                                        scheduler=None)

In [ ]:
cfeats.stack('channel').iloc[:24]

We can see that for some objects, there appears to be a common frequency / period for all its observed passbands, whereas for others the calculated frequencies are all different and seemingly random. We may assume that the objects with completely mismatching frequencies or very low frequencies (very long periods) actually do not exhibit periodic behaviour. Let us look at a few examples where the periods do match up between most bands:

In [ ]:
def plot_phase(n, fr):
    selected_times = times_list[n]
    selected_flux = flux_list[n]
    colors = ['red', 'orange', 'yellow', 'green', 'blue', 'purple']
    f, ax = plt.subplots(figsize=(12, 6))
    for band in range(6):
        ax.scatter(x=(selected_times[band] * fr) % 1, 
                   y=selected_flux[band], 
                   c=colors[band])
    ax.set_xlabel('phase')
    ax.set_ylabel('relative flux')
    ax.set_title(
        f'object {train_metadata["object_id"][n]}, class {train_metadata["target"][n]}')
    plt.show()

In [ ]:
plot_phase(0, 3.081631)

In [ ]:
plot_phase(3, 0.001921)

In [ ]:
plot_phase(6, 1.005547)

For periodic objects, observations made in the same phase should have similar relative flux values. As we see, Object 615 is clearly periodically changing its magnitude,  whereas object 745 and 1598's 'periodicity' are more like staying stationary most of the time with sudden bursts.

We kind of found a pattern here. If all colour bands agree in frequency, an object is likely to have periodic behaviours. When the colour bands have wildly different frequencies, the object is likely aperiodic. When most bands have very small frequencies (longer period than the observation window) and a few bands have larger detected frequencies, it is likely a burst event.

It appears that calculated periods alone cannot tell us the full picture, as there are plenty of objects in the dataset without clear periodic patterns. However, determining whether an object has a periodic pattern can be very important in determining the object class, and for objects that are actually periodic, the time series in phase space can tell us a lot more than in time space. Therefore, for a large portion of the dataset, conversion to phase space is actually a very powerful preprocessing step to use.

We will plot below a few more examples of where phase conversion works well:

In [ ]:
plot_phase(46, 1.781711)

In [ ]:
plot_phase(62, 4.771011)

In [ ]:
plot_phase(69, 1.597659)

In [ ]:
plot_phase(91, 2.668811)

It appears that we are onto something. The classes 16 and 92 appear to be quite periodic where class 90 is what appears to be a group of burst events. So by just calculating a few frequency features and comparing these features between bands, we can already roughly tell them apart. We might still have trouble distinguishing between class 16 and 92, but with frequency divided out through phase-space conversion, we essentially have higher density data for these two classes to work with, and building a model on top of that should be easier.

One more thing... We have to figure out if the same procedure can work well on non-DDF objects.

In [ ]:
nonddf_pos = train_metadata[train_metadata['ddf'] == 0].index
nonddf_times_list = [v for i, v in enumerate(times_list) if i in set(nonddf_pos)]
nonddf_flux_list = [v for i, v in enumerate(flux_list) if i in set(nonddf_pos)]

In [ ]:
warnings.simplefilter('ignore', RuntimeWarning)
N = 50
cfeats = featurize.featurize_time_series(times=nonddf_times_list[:N],
                                        values=nonddf_flux_list[:N],
                                        features_to_use=['freq1_freq',
                                                        'freq1_signif',
                                                        'freq1_amplitude1'],
                                        scheduler=None)

In [ ]:
cfeats.stack('channel').iloc[:24]

In [ ]:
plot_phase(nonddf_pos[15], 4.440506)

In [ ]:
plot_phase(nonddf_pos[20], 0.831637)

In [ ]:
plot_phase(nonddf_pos[39], 0.817696)

In [ ]:
plot_phase(nonddf_pos[46], 2.514430)

Generally, we found that we are still able to use frequency analysis to find the most apparent periodic classes like 16 and 92, but we have to be much more lenient on how different bands agree with each other (sometimes only 3-4 bands will strongly agree in terms of frequency), and we might have to pay much more attention to the significance level of these frequencies. This could be evidence to support the case that DDF and non-DDF objects should be dealt with using different models, or at least treated differently in model.